diag_manager/diag_grid.F90

```diff
@@ -132,10 +132,10 @@ MODULE diag_grid_mod
   !! and before the first call to register the fields.
   SUBROUTINE diag_grid_init(domain, glo_lat, glo_lon, aglo_lat, aglo_lon)
     TYPE(domain2d), INTENT(in) :: domain !< The domain to which the grid data corresponds.
-    REAL, INTENT(in), DIMENSION(:,:) :: glo_lat !< The latitude information for the grid tile.
-    REAL, INTENT(in), DIMENSION(:,:) :: glo_lon !< The longitude information for the grid tile.
-    REAL, INTENT(in), DIMENSION(:,:) :: aglo_lat !< The latitude information for the a-grid tile.
-    REAL, INTENT(in), DIMENSION(:,:) :: aglo_lon !< The longitude information for the a-grid tile.
+    CLASS(*), INTENT(in), DIMENSION(:,:) :: glo_lat !< The latitude information for the grid tile.
+    CLASS(*), INTENT(in), DIMENSION(:,:) :: glo_lon !< The longitude information for the grid tile.
+    CLASS(*), INTENT(in), DIMENSION(:,:) :: aglo_lat !< The latitude information for the a-grid tile.
+    CLASS(*), INTENT(in), DIMENSION(:,:) :: aglo_lon !< The longitude information for the a-grid tile.
```

Comment:
Two comments/questions:
1. Is checking if they of the same real kind necessary (see the code snippet below)? Or is it sufficient to know that all real's coming from the dycore are of the same kind, and those from the mom6 are of the same kind?
```
class(*), intent(in) :: a, b, c, d

select type (a) 
type is (real(4))
  select type (b) 
  type is (real(4))
    select type (c) 
    type is (real(4))
      select type (d) 
      type is (real(4))
      type is (real(8))
        ! ERROR
      end select
    type is (real(8))
      ! ERROR
    end select
  type is (real(8))
    ! ERROR
  end select
type is (real(8))
  select type (b) 
  type is (real(4))
    ! ERROR
  type is (real(8))
    select type (c) 
    type is (real(4))
      ! ERROR
    type is (real(8))
      select type (d) 
      type is (real(4))
        ! ERROR
      type is (real(8))
      end select
    end select
  end select
end select
```

2. Is checking class are real types somewhat redundant to adding class default in your comment below? For example,
```
class(*), intent(in) :: a

! Checking a is of real kind
select type (a)
type is (real(4))
type is (real(8))
class default
  ! ERROR
end select

! Actual use of a
select type (a)
type is (real(4))
  print*, a
type is (real(8))
  print*, a
class default  ! <--- Isn't this redundant
  ! ERROR
end select
```

Reply:
Can we just do
1) initial check on each variable separately
2) assume all variables are of the same real type
3) omit `class default` when the variables are used?
```
class(*), intent(in) :: a, b, c, d

! Initial check on each variable separately
select type (a)
type is (real(4))
type is (real(8))
class default
  ! ERROR
end select

select type (b)
type is (real(4))
type is (real(8))
class default
  ! ERROR
end select

! same check for c and d

! Use variables assuming they are of the same kind (they are already known to be real at this point)
select type (a)
type is (real(4))
  select type (b)
  type is (real(4))
    select type (c)
    type is (real(4))
      select type (d)
      type is (real(4))
        ! DO SOMETHING
      end select
    end select
  end select
type is (real(8))
  select type (b)
  type is (real(8))
    select type (c)
    type is (real(8))
      select type (d)
      type is (real(8))
        ! DO SOMETHING
      end select
    end select
  end select
end select
```

Reply:
For 1, since these changes will eventually be going into the main code, it's best to cover other use cases as well so you can't assume they'd always be the same kinds.
I know the syntax makes this a bit tedious, but setting a logical value whenever you know the types are valid can help with redundant checks. From your second comment, you can remove the individual checks if you just set a logical when they are used:
```
class(*), intent(in) :: a, b, c, d
logical :: valid_types = .false.

select type (a)
type is (real(4))
  select type (b)
  type is (real(4))
    select type (c)
    type is (real(4))
      select type (d)
      type is (real(4))
        ! DO SOMETHING
        valid_types = .true.
      end select
    end select
  end select
type is (real(8))
  select type (b)
  type is (real(8))
    select type (c)
    type is (real(8))
      select type (d)
      type is (real(8))
        ! DO SOMETHING
        valid_types = .true.
      end select
    end select
  end select
end select
if(.not. valid_types) ! ERROR
```
This way the code can't fail silently if it doesn't reach the `! DO SOMETHING` line.

For 2, you are correct, that would be redundant. There only really needs to be one class default/error for an argument so it can just be in the first use rather than adding an additional `select type`.